Source/Core/Core/IOS/USB/Emulated/Microphone.cpp

```diff
@@ -0,0 +1,371 @@
+
+    cubeb_stream_params params{};
+    params.format = CUBEB_SAMPLE_S16LE;
+    params.rate = sampling_rate;
+    params.channels = 1;
+    params.layout = CUBEB_LAYOUT_MONO;
+
+    u32 minimum_latency;
+    if (cubeb_get_min_latency(m_cubeb_ctx.get(), &params, &minimum_latency) != CUBEB_OK)
+    {
+      WARN_LOG_FMT(IOS_USB, "Error getting minimum latency");
+      minimum_latency = 16;
+    }
+
+    cubeb_devid input_device =
+        CubebUtils::GetInputDeviceById(Config::Get(Config::MAIN_WII_SPEAK_MICROPHONE));
+    if (cubeb_stream_init(m_cubeb_ctx.get(), &m_cubeb_stream, "Dolphin Emulated Wii Speak",
+                          input_device, &params, nullptr, nullptr,
+                          std::max<u32>(16, minimum_latency), DataCallback, StateCallback,
+                          this) != CUBEB_OK)
+    {
+      ERROR_LOG_FMT(IOS_USB, "Error initializing cubeb stream");
+      return;
+    }
+
+    if (cubeb_stream_start(m_cubeb_stream) != CUBEB_OK)
+    {
+      ERROR_LOG_FMT(IOS_USB, "Error starting cubeb stream");
+      return;
+    }
+
+    INFO_LOG_FMT(IOS_USB, "started cubeb stream");
+  });
+}
+
+void Microphone::StreamStop()
+{
+  if (!m_cubeb_stream)
+    return;
+
+  m_worker.Execute([this] {
+    if (cubeb_stream_stop(m_cubeb_stream) != CUBEB_OK)
+      ERROR_LOG_FMT(IOS_USB, "Error stopping cubeb stream");
+    cubeb_stream_destroy(m_cubeb_stream);
+    m_cubeb_stream = nullptr;
+  });
+}
+
+long Microphone::DataCallback(cubeb_stream* stream, void* user_data, const void* input_buffer,
+                              void* /*output_buffer*/, long nframes)
+{
+  // Skip data when core isn't running
+  if (Core::GetState(Core::System::GetInstance()) != Core::State::Running)
+    return nframes;
+
+  // Skip data when HLE Wii Speak is muted
+  // TODO: Update cubeb and use cubeb_stream_set_input_mute
+  if (Config::Get(Config::MAIN_WII_SPEAK_MUTED))
+    return nframes;
+
+  auto* mic = static_cast<Microphone*>(user_data);
+  const auto& sampler = mic->GetSampler();
+
+  // Skip data if sampling is off or mute is on
+  if (!sampler.sample_on || sampler.mute)
+    return nframes;
+
+  std::lock_guard lk(mic->m_ring_lock);
+  std::span<const s16> buff_in(static_cast<const s16*>(input_buffer), nframes);
+  const auto gain = mic->ComputeGain(Config::Get(Config::MAIN_WII_SPEAK_VOLUME_MODIFIER));
+  auto processed_buff_in = buff_in | std::views::transform([gain](s16 sample) {
+                             return MathUtil::SaturatingCast<s16>(sample * gain);
+                           });
+
+  mic->UpdateLoudness(processed_buff_in);
+
+  for (s16 le_sample : processed_buff_in)
+  {
+    mic->m_stream_buffer[mic->m_stream_wpos] = Common::swap16(le_sample);
+    mic->m_stream_wpos = (mic->m_stream_wpos + 1) % mic->STREAM_SIZE;
+  }
+
+  mic->m_samples_avail += nframes;
+  if (mic->m_samples_avail > mic->STREAM_SIZE)
+  {
+    WARN_LOG_FMT(IOS_USB, "Wii Speak ring buffer is full, data will be lost!");
+    mic->m_samples_avail = 0;
```

Comment:
Similar situation as with `m_loudness_level`. `m_samples_avail` is read from core-thread unprotected, but it's maybe a bit more complicated with the buffer-overflow logic here.

Does setting `m_samples_avail` to zero like this even make sense? Maybe it should just be capped to `STREAM_SIZE` ? I think it can then just be atomic.

Reply:
Not sure how to make it safe with the ring buffer logic so I added a `std::lock_guard` and marked the `std::mutex` as `mutable`.

Regarding `m_samples_avail = 0` I was hoping that it will let the CPU catch up. However, as you pointed out, it doesn't make much sense. If the CPU is catching up, it will catch up with the buffer being full regardless.